docs/design/generic.md

```diff
@@ -0,0 +1,95 @@
+# Generic artifact fetching
+
+## Introduction
+
+This document will describe high-level implementation overview for supporting generic artifact fetching in cachi2.
+Up until now cachi2 has only supported package managers for various ecosystems and languages.
+However, there are a couple of use-cases where language non-specific artifacts need to be pre-fetched in order to satisfy
+requirements of a hermetic build.
+
+## Context
+
+For context, generic artifact fetching is a use-case of its own (e.g. [OVAL feeds](https://github.com/CISecurity/OVALRepo),
+AI models), it is also necessary precursor for implementing support for fetching maven artifacts, which won't be covered
+in this design, but in a followup document.
+
+## Design
+
+In this section, I will try to cover individual parts of the design.
+
+### Source repository
+
+This section will describe the structure of the source repository, that will serve as an input to cachi2. The idea is to
+define a cachi2 lockfile that will specify individual artifacts to fetch along with necessary metadata - e.g. checksums.
+The format chosen for this lockfile is yaml, and will include [purl](https://github.com/package-url/purl-spec) for each
+of the fetched artifacts. This decision was made mainly because it allows for followup implementation of maven support,
+with accurate SBOM information. Here's an example of such a lockfile.
+
+```yaml
+artifacts:
+  - purl: pkg:generic/granite-model?download_url=https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00001-of-00003.safetensors?download=true
+    target: granite-model.safetensors
+    checksums:
+      sha256: 07123e1f482356c415f684407a3b8723e10b2cbbc0b8fcd6282c49d37c9c1abc
+```
+
+#### Lockfile format and validation
+
+##### purl (required)
```

Comment:
I wonder if purls provide the best user experience here. Maybe we could allow either:
- a pair of `download_url` and `checksums`
- a purl containing the info we need (which happens to be the fields above)

I believe this lockfile will be written manually in many cases, so I think keeping the user experience here is worth it.

Reply:
@eskultety I remember you had a point against having a Cachi2-defined lockfile format?

Reply:
Also, wdyt about asking the user to specify the checksum as part of a purl qualifier (since it is officially supported)?

Reply:
And do you think it's worth allowing users to just set `download_url` and `checksums`, as an alternative input format?

Reply:
I even wonder if this purl input format should be exclusive to `maven-artifacts` :thinking:

Reply:
I think allowing both is actually a great idea, if they are mutually exclusive. 

**download_url + checksums:**
- more readable, easier to manually specify
- will ever only produce a generic purl
 
**purl:**
- less readable, more of a mess to manually specify
- chance of producing quality sbom for maven pkgs 

On the exclusiveness of purl to `pkg:maven`, I honestly can't decide. Maybe leaning more towards no, because if the lockfile is going to be generated automatically, you'd maybe want to have a single format for everything. Not sure how hard of an argument that is.

Reply:
I'm ok with having both options, or just keeping the "download_url + checksums" one for now.

Reply:
Let's do both. I'll update this in the design.

Reply:
With https://github.com/containerbuildsystem/cachi2/pull/663 in mind (which was opened 2 days ago I lacked any insight into Maven handling) I think we should drop modeling PURL as part of this particular design doc, and especially `pkg:generic`, and keep it simple. In context of Maven `generic` IMO just looks confusing. Maven can enhance the format with a particular PURL spec type for the use case. We could adopt `pkg:generic` for input when we have a use case, otherwise we'd similarly to my original CycloneDX proposal be addressing an immediately non-existent use case.

Should be a small and straight forward change in this design, shouldn't block anything.

Reply:
I'm not sure I completely understand, let me see if I got it right. Are you saying that you only want to accept the `download_url + checksums` for the generic fetcher part, and add an option to specify `purl`, restricted to only `pkg:maven` in #663?

Reply:
Yes.

Reply:
Fixed, good point.